scripts/irods/distro_distill.py

```diff
@@ -0,0 +1,323 @@
+def package_type():
+	global _cached_package_type
+	if _cached_package_type is not None:
+		return _cached_package_type.value
+	log = logging.getLogger(__name__)
+	distro_type = distribution_type()
+	if distro_type in ['debian', 'ubuntu']:
+		pkg_type = 'deb'
+	elif distro_type in ['amzn', 'fedora', 'rhel', 'scientific', 'sles', 'sled', 'suse']:
+		pkg_type = 'rpm'
+	elif distro_type in ['alpine']:
+		pkg_type = 'apk'
+	elif distro_type in ['arch']:
+		pkg_type = 'pacman'
+	else:
+		module_fname = Path(__file__).name
+		log.warning(f'{module_fname} does not know what package type is used in distro \'{distro_type}\'.')
+		pkg_type = None
+	_cached_package_type = _CachedValue(pkg_type)
+	return pkg_type
+_cached_package_type = None
+
+# Returns the filename extension of distribution packages
+# NOTE: If we need to add proper support for platforms that have native packages with
+# multiple filename extensions (arch, for example), this will need some tweaking
+def package_filename_extension():
+	global _cached_package_filename_extension
+	if _cached_package_filename_extension is not None:
+		return _cached_package_filename_extension.value
+	log = logging.getLogger(__name__)
+	pkg_type = package_type()
+	if pkg_type == 'deb':
+		pkg_suffix = 'deb'
+	elif pkg_type == 'rpm':
+		pkg_suffix = 'rpm'
+	elif pkg_type == 'apk':
+		pkg_suffix = 'apk'
+	elif pkg_type == 'pacman':
+		pkg_suffix = 'pkg.tar'
+	elif pkg_type is None:
+		log.warning('Package type must be known before package filename extension can be discerned.')
+		pkg_suffix = None
+	else:
+		module_fname = Path(__file__).name
+		log.warning(f'{module_fname} does not know what package filename extension is used for package type \'{pkg_type}\'.')
+		pkg_suffix = None
+	_cached_package_filename_extension = _CachedValue(pkg_suffix)
+	return pkg_suffix
+_cached_package_filename_extension = None
+
+# Returns the architecture string to use for packaging
+def package_architecture_string():
+	global _cached_package_architecture_string
+	if _cached_package_architecture_string is not None:
+		return _cached_package_architecture_string.value
+	log = logging.getLogger(__name__)
+	if package_type() == 'deb':
+		encoding = locale.getpreferredencoding(False)
+		pkg_arch = subprocess.check_output(['dpkg', '--print-architecture'], timeout=5, encoding=encoding)
+		pkg_arch = pkg_arch.strip()
+	else:
+		if package_type() is None:
+			log.warning('Unknown package type, architecture string may be inaccurate.')
+		pkg_arch = platform.machine()
+	_cached_package_architecture_string = _CachedValue(pkg_arch)
+	return pkg_arch
+_cached_package_architecture_string = None
+
+
+# This implementation of main is written specifically for use by the irods buildsystem, and will be changed
+# significantly for the standalone release of this tool.
+def main():
+	parser = argparse.ArgumentParser(description='distro info tool', add_help=True)
+	data_group = parser.add_mutually_exclusive_group(required=True)
+	data_group.add_argument(
```

Comment:
Seems fine.... I would request a little bit more detail about what this script does, either by an appropriate name change or by adding argument help text.

Reply:
Oh, I see now that some sample uses are in  the `cmake/Modules/IrodsCPackPlatform.cmake` file, so that part's fine. I think  it might still be a good thing however to document a couple of the mechanisms that are key to what it's doing, for maintainers, possibly in doc-strings or in comments strewn internally in some of the functions.

Reply:
Better documentation, help text, and more are planned for the standalone package, which will replace this implementation and the one in externals.

Reply:
@d-w-moore Are there specific places in the function implementations where you feel some additional wording would be helpful? If so, please highlight them.

Reply:
Will do.

Reply:
In general, and it may be a programming philosophy choice, but I personally prefer to define return values early in a function, preferably initialized to a value that is a useful default or some zeroed/meaningless value. That would certainly give the reader less to do, in terms of the mental gymnastics of running the code in their heads ; )

Reply:
The comments at the head of each function seem enough, with that in mind

Reply:
That's a little difficult to do at the moment due to the current approach to return value caching, but one of the plans for the standalone package is to put everything in a class the same way `distro` does, so we will be able to use `functools.cached_property` (or a shim if running Python < 3.9) and all of that will go away. Additionally, type hints are also planned, which should make things a bit easier to follow.

Reply:
Sounds like you're in agreement that the current impl is good enough.

If so, please resolve this conversation.

@SwooshyCueb Do we have an issue for the "future" work you're planning?

Reply:
Yes, #8146, though it doesn't have all my plans/notes in it currently